### PR TITLE
Dynamic bitset

### PR DESCRIPTION
**Pull request by JohanMabille**
based on #16 , partially replaces #13 

You can ignore the first 3 commits for the review (they will disappear when #16 is merged and this one is rebased).

### CODE REVIEW

File: include/sparrow/dynamic_bitset.hpp [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    public:
+
+        using base_type = dynamic_bitset_base<buffer<T>>;
+        using storage_type = typename base_type::storage_type;
+        using block_type = typename base_type::block_type;
+        using value_type = typename base_type::size_type;
+        using size_type = typename base_type::size_type;
+
+        dynamic_bitset();
+        explicit dynamic_bitset(size_type n, value_type v = false);
```
Comment (jjerphan) [inline include/sparrow/dynamic_bitset.hpp:108]: Reminder of one of @Klaim's comment: two constructors need to be defined.

File: include/sparrow/dynamic_bitset.hpp [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        dynamic_bitset_base(storage_type&& buffer, size_type size, size_type null_count);
+        ~dynamic_bitset_base() = default;
+
+        dynamic_bitset_base(const dynamic_bitset_base&) = default;
+        dynamic_bitset_base(dynamic_bitset_base&&) = default;
+
+        dynamic_bitset_base& operator=(const dynamic_bitset_base&) = default;
+        dynamic_bitset_base& operator=(dynamic_bitset_base&&) = default;
+
+        void resize(size_type n, value_type = false);
```
Comment (jjerphan) [inline include/sparrow/dynamic_bitset.hpp:69]:
```suggestion
        void resize(size_type n, value_type b = false);
```

File: include/sparrow/dynamic_bitset.hpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    template <class B>
+    void dynamic_bitset_base<B>::set(size_type pos, value_type value)
+    {
+        assert(pos < size());
+        block_type& block = m_buffer.data()[block_index(pos)];
+        const bool old_value = block & bit_mask(pos);
+        if (value)
+        {
+            block |= bit_mask(pos);
+            if (m_null_count && !old_value)
```
Comment (jjerphan) [inline include/sparrow/dynamic_bitset.hpp:182]:
```suggestion
            if (!old_value)
```

File: include/sparrow/dynamic_bitset.hpp [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+     */
+    template <class T>
+    class dynamic_bitset : public dynamic_bitset_base<buffer<T>>
+    {
+    public:
+
+        using base_type = dynamic_bitset_base<buffer<T>>;
+        using storage_type = typename base_type::storage_type;
+        using block_type = typename base_type::block_type;
+        using value_type = typename base_type::size_type;
```
Comment (jjerphan) [inline include/sparrow/dynamic_bitset.hpp:104]:
```suggestion
        using value_type = typename base_type::value_type;
```

**Review by jjerphan — COMMENTED** [review]
A first pass.

File: test/test_dynamic_bitset.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            CHECK_EQ(bm.data()[0], 46);
+            CHECK_EQ(bm.null_count(), m_null_count - 1);
+
+            bm.set(24, 0);
+            CHECK_EQ(bm.data()[3], 6);
+            CHECK_EQ(bm.null_count(), m_null_count);
+
+            bm.set(24, 0);
+            CHECK_EQ(bm.data()[3], 6);
+            CHECK_EQ(bm.null_count(), m_null_count);
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply test/test_dynamic_bitset.cpp:159]: No, I'm checking that setting the same value does not wrongly update the null count.

File: test/test_dynamic_bitset.cpp [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        TEST_CASE_FIXTURE(bitmap_fixture, "move semantic")
+        {
+            bitmap bref(p_buffer, m_size);
+            bitmap b(bref);
+
+            bitmap b2(std::move(b));
+            CHECK_EQ(b2.size(), bref.size());
+            CHECK_EQ(b2.null_count(), bref.null_count());
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply test/test_dynamic_bitset.cpp:117]: nope, the pointer are different, the value have been copied.

File: test/test_dynamic_bitset.cpp [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            const std::size_t expected_block_count = 2;
+            std::uint8_t* buf = new std::uint8_t[expected_block_count];
+            buf[0] = 37;
+            buf[1] = 2;
+            bitmap b4(buf, expected_block_count * 8);
+            bitmap b5(b4);
+
+            b2 = std::move(b4);
+            CHECK_EQ(b2.size(), b5.size());
+            CHECK_EQ(b2.null_count(), b5.null_count());
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply test/test_dynamic_bitset.cpp:132]: Same here.

**Review by Klaim — APPROVED** [review]
LGTM

**Pull request merged**